Sources/MapboxCoreNavigation/CacheHandlerFactory.swift

```diff
@@ -0,0 +1,73 @@
+import MapboxDirections
+import MapboxNavigationNative
+
+protocol CacheHandlerData {
+    var tileStorePath: String { get }
+    var credentials: DirectionsCredentials { get }
+    var tilesVersion: String { get }
+    var historyDirectoryURL: URL? { get }
+    var targetVersion: String? { get }
+    var configFactoryType: ConfigFactory.Type { get }
+}
+
+extension NativeHandlersFactory: CacheHandlerData { }
+
+enum CacheHandlerFactory {
+    
+    private struct CacheKey: CacheHandlerData {
+        let tileStorePath: String
+        let credentials: DirectionsCredentials
+        let tilesVersion: String
+        let historyDirectoryURL: URL?
+        let targetVersion: String?
+        let configFactoryType: ConfigFactory.Type
+        
+        init(data: CacheHandlerData) {
+            self.tileStorePath = data.tileStorePath
+            self.credentials = data.credentials
+            self.tilesVersion = data.tilesVersion
+            self.historyDirectoryURL = data.historyDirectoryURL
+            self.targetVersion = data.targetVersion
+            self.configFactoryType = data.configFactoryType
+        }
+        
+        private static func optionalsAreNotEqual<T: Comparable>(_ lhs: T?, _ rhs: T?) -> Bool{
+            if let firstVal = lhs, let secondVal = rhs {
+                return firstVal != secondVal
+            }
+            else {
+                return lhs != nil || rhs != nil
+           }
+        }
+        
+        static func != (lhs: CacheKey, rhs: CacheHandlerData) -> Bool {
```

Comment:
I am not sure why I needed `optionalsAreNotEqual` exactly, probably a leftover from testing [mis]matching keys. `Equatable` protocol is not ideal here because we need to compare different types (`CacheKey` vs. `CacheHandlerData`). Extending `CacheHandlerData` with `Equatable` is not possible because it introduces a dependency to `Self` type and won't help comparing different types.